### PR TITLE
Pokemon: Reign in triggering; Trigger for accepted pokemon names only

### LINKED ISSUE #2833: Pokemon: Reign in triggering; Check query against list of Pokemon Names
This IA is over triggering a **_lot_**, however thankfully it is not returning irrelevant result.

We should still avoid making unnecessary API calls and we have a simple way to handle it in this case.

We should copy the list of accepted Pokemon (available [here](https://github.com/PokeAPI/pokeapi/blob/master/data/v2/csv/pokemon.csv)) and use it to lookup the candidate Pokemon name. If it matches we should move forward with the API call. If not we should `return`.

------
IA Page:  http://duck.co/ia/view/pokemon_data 
[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @AkA84

### PR DESCRIPTION
**Pull request by sahildua2305**
###### What does your Pull Request do (check all that apply)?

- [ ] New Instant Answer: **`New {IA Name} Instant Answer`**
- [x] Improvement
    - [x] Bug fix: **`Pokemon: Fix #2833`**
    - [ ] Enhancement: **`{IA Name}: {Description of Improvements}`**
- [ ] Non–Instant Answer
    - [ ] Other (Role, Template, Test, Documentation, etc.): **`{SpiceRole/Templates/Tests/Docs}: {Short Description}`**

###### Description of changes
Downloaded the list of acceptable Pokemon names and added a check against this list before sending API call.

###### Which issues (if any) does this fix?

Fixes #2833 

###### People to notify (@mention interested parties)
@moollaza 

---

Instant Answer Page: https://duck.co/ia/view/pokemon_data

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @AkA84

### CODE REVIEW

**Comment by daxtheduck** [comment]
## [Pokemon](https://duck.co/ia/view/pokemon_data)

**Description**: Returns Pokémon data from the Pokéapi API

**Example Query**: [pokemon charizard](https://beta.duckduckgo.com/?q=pokemon charizard), [bulbasaur pokemon](https://beta.duckduckgo.com/?q=bulbasaur pokemon)

**Tab Name**: Pokedex

**Source**:  

*These are the important fields from the IA page.  Please check these for errors or missing information and update the [IA page](https://duck.co/ia/view/pokemon_data)*

---
*This is an automated message which will be updated as changes are made to the [IA page](https://duck.co/ia/view/pokemon_data)*

File: lib/DDG/Spice/Pokemon/Data.pm [hunk 1]
```diff
@@ -21,9 +21,11 @@ spice alt_to => {
 	}
 };
 
+my $accepted_names = join "|", share('pokemon-names.txt')->slurp(chomp => 1);
+
 # Handle statement
 handle remainder => sub {
-    return lc $_ unless !$_ or $_ =~ /\s/;
+    return lc $_ unless !$_ or $_ =~ /\s/ or !($_ =~ /\b($accepted_names)\b/i);
```
Comment (moollaza) [inline lib/DDG/Spice/Pokemon/Data.pm]: I would do:

```perl
my %pokemon = map { trim($_) => 0 } share('pokemon-names.txt')->slurp;	

handle remainder_lc => sub {
return unless $_;
return unless exists $pokemon{$_};
return $_;
```

**Comment by moollaza** [comment]
@sahildua2305 this is great! Please add some tests that confirm incorrect queries don't trigger this. E.g. "Pokemon game", "Pokemon movie", etc

More examples:

drake_at_duckduckgo

File: lib/DDG/Spice/Pokemon/Data.pm (hunk 1 above)
Comment (sahildua2305) [inline lib/DDG/Spice/Pokemon/Data.pm]: @moollaza Again, I am new to Perl (still learning). Thank you so much for introducing me to new & efficient ways to do things. :smile:

Comment (sahildua2305) [inline lib/DDG/Spice/Pokemon/Data.pm]: @moollaza I just realized that this solution will be case sensitive. It was easy to make the previous solution case insensitive.

How can I make this solution case insensitive? :confused:

Comment (moollaza) [inline lib/DDG/Spice/Pokemon/Data.pm]: @sahildua2305 this _**is**_ case insensitive 👍 

`handle remainder_lc` forces the query input to be lowercased. so `Pokemon CHARIZARD` would have `$_ == "charizard"`, and "charizard" exists in our hash :)

Comment (sahildua2305) [inline lib/DDG/Spice/Pokemon/Data.pm]: Yes! I just figured it out. Thanks @moollaza

**Comment by sahildua2305** [comment]
@moollaza I have updated the code.

Please note that it needs a new dependency - `String::Util`.

File: lib/DDG/Spice/Pokemon/Data.pm [hunk 2]
```diff
@@ -2,6 +2,7 @@ package DDG::Spice::Pokemon::Data;
 # ABSTRACT: Returns pokemon data from the pokeapi API
 
 use DDG::Spice;
+use String::Util qw(trim);
```
Comment (moollaza) [inline lib/DDG/Spice/Pokemon/Data.pm]: @sahildua2305 please switch to `use Text::Trim;` -- we've got that one in production already 👍

File: lib/DDG/Spice/Pokemon/Data.pm [hunk 3]
```diff
@@ -21,10 +22,13 @@ spice alt_to => {
...
+my %pokemon = map { trim($_) => 0 } share('pokemon-names.txt')->slurp;
+
 # Handle statement
-handle remainder => sub {
-    return lc $_ unless !$_ or $_ =~ /\s/;
-    return;
+handle remainder_lc => sub {
+    return unless $_;
+    return unless exists $pokemon{$_};
+    return lc $_;
```
Comment (moollaza) [inline lib/DDG/Spice/Pokemon/Data.pm]: You don't need `lc` here. `remainder_lc` forces `$_` to be lowercased already 👍

**Comment by sahildua2305** [comment]
@moollaza Please check :smile:

**Comment by sahildua2305** [comment]
@moollaza Can we go ahead with this one?

**Comment by moollaza** [comment]
@sahildua2305 Thanks! This LGTM :+1:

**Pull request merged**